### PR TITLE
Add keadm-debug proposal

### LINKED ISSUE #1918: keadm get : Get and format the data in the output database edgecore.db
**What would you like to be added/modified**:
Get and format the data in the output database edgecore.db like `kubectl get`, help debug.
**Why is this needed**:
Various types of data now are stored in edge-side database **edgecore.db**(/var/lib/kubeedge/edgecore.db)  such as :
- pod
- secret
- configmap
- service
- endpoints
- node

Now, If we want to confirm whether a specific resource is successfully delivered to the edge, we need to use `sqlite3 /var/lib/kubeedge/edgecore.db` to get the data in edgecore.db, as below:
image

It is troublesome and messy comparing to `kubectl get`:
image
image

Tips:
1. How to get data in edgecore.db?
Refer to `edge\pkg\metamanager\dao\meta.go`

2. How to add a command to keadm?
Refer to `keadm\cmd\keadm\app\cmd\edge\join.go`

3. How to fomat data?
Refer to `kubectl get`

### PR DESCRIPTION
**Pull request by qingchen1203**
**What type of PR is this?**
/kind design
/kind feature

**What this PR does / why we need it**:
Now, If we want to confirm whether a specific resource is successfully delivered to the edge, we need to use sqlite3 /var/lib/kubeedge/edgecore.db to get the data in edgecore.db. This way is very inconvenient.
So, we need an edge-side tool like kubectl for system check, data format display and fault detection.

**Which issue(s) this PR fixes**:

Fixes #1918

**Special notes for your reviewer**:
Under writing

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: docs/proposals/keadm-debug.md [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+`keadm debug get`
+
+- What is it?
+  
+  - This command will get and format the specified resource`s information from the local database of the edge node
+  
+- What shall be its scope ?
+
+  1. Format resource information from the local database, and available resource types:
+    - `pod`
```
Comment (GsssC) [inline docs/proposals/keadm-debug.md:307]:
```suggestion
     - `all`
     - `pod`
```

File: docs/proposals/keadm-debug.md [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+keadm help command provide debug function to help diagnose the cluster
+
+Usage:
+  keadm debug [command]
+
+Examples:
+
+Available Commands:
+  diagnose        Diagnose specific fault scenarios in an all-round way and locate the cause of the fault
+  collect         Obtain all data of the current node, and then locate and use operation personnel
```
Comment (GsssC) [inline docs/proposals/keadm-debug.md:53]: what dose it mean?

File: docs/proposals/keadm-debug.md [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+- What shall be its scope ?
+
+  1. Format resource information from the local database, and available resource types:
+    - `pod`
+    - `node`
+    - `service`
+    - `secret`
+    - `configmap`
+    - `endpoint`
+  2. Use flag `-n, --namespace=''` to indicate the scope of resource acquisition, if the flag `-A, --all-namespaces` is used, information of the specified resource will be obtained from all ranges
```
Comment (GsssC) [inline docs/proposals/keadm-debug.md:313]: 1. add more options such as edgecore path that will be visited, default to `/var/lib/kubeedge/edgecore.db`
2. detailing the  option `-o` ,what formats we support?
- yaml
- json
- **wide**

File: docs/proposals/keadm-debug.md [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+    
+   - armv7i (arm32) architecture
+       Raspbian GNU/Linux 9 (stretch)
+    
+   - aarch64 (arm64) architecture
+       Ubuntu 18.04.2 LTS (Bionic Beaver)
+  3. Use command `cpu` can determine if the NUMBER of CPU cores meets the requirement, minimum 1Vcores
+  4. Use command `memory` check the system memory size, and the amount of memory left, requirements minimum 256MB
+  5. Use command `disk` check whether the disk meets the requirements, requirements minimum 1 GB
+  6. Use command `dns` Check whether the node domain name resolution function is normal
```
Comment (GsssC) [inline docs/proposals/keadm-debug.md:215]: word `function` may be redundant

File: docs/proposals/keadm-debug.md [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+  7. Use command `runtime `  Check whether the node container runtime function is installed, can use parameter `--runtime` to set container runtime,  default is docker
```
Comment (GsssC) [inline docs/proposals/keadm-debug.md:216]: ditto. word `function`.
and may only support docker in alpha.

File: docs/proposals/keadm-debug.md [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+  8. Use command `network `  check whether the node can communicate with the endpoint on the cloud,    can use parameter `--ip` to set test ip, default to ping clusterdns
```
Comment (GsssC) [inline docs/proposals/keadm-debug.md:217]: is not to ping cloudcore ip? there may be not clusterdns at edge.

File: docs/proposals/keadm-debug.md [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+
+Many users shared their feedback that kubeEdge edge nodes do not have good debugging and diagnostic methods, which may prevent people from trying kubeEdge.
+There should be a simple and clear way to help operation and maintenance personnel to ensure the stable operation of kubeEdge, so that users can focus more on using it immediately.
+
+# Proposal
+
+KubeEdge should have simple commands to debug and troubleshoot edge components.
+Therefore, it is recommended to use the following commands during operation and maintenance of KubeEdge.
+
+## Inscope
```
Comment (GsssC) [inline docs/proposals/keadm-debug.md:23]: Add the release plan such as:
- alpha
- beta
- ga

File: docs/proposals/keadm-debug.md [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+  
+- What shall be its scope ?
+    1. Use command `all` can diagnose all resource
+    2. Use command `node` can troubleshoot the cause of edge node failure with installed software
+       1. check system resources is enough
+       2. check container runtime is running 
+       3. check all edgecore components are running
+       4. check cloudercore can be connected
+    3. Use command `pod` can troubleshooting specific container application instances on nodes
+       1. check pod Is the configuration correct
```
Comment (GsssC) [inline docs/proposals/keadm-debug.md:186]: It is not a good place to check pod configuration.

File: docs/proposals/keadm-debug.md [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+    1. Use command `all` can diagnose all resource
+    2. Use command `node` can troubleshoot the cause of edge node failure with installed software
+       1. check system resources is enough
+       2. check container runtime is running 
+       3. check all edgecore components are running
+       4. check cloudercore can be connected
+    3. Use command `pod` can troubleshooting specific container application instances on nodes
+       1. check pod Is the configuration correct 
+       2. check pod  image can be right pull
+       3. check pod schedule
```
Comment (GsssC) [inline docs/proposals/keadm-debug.md:188]: ditto. not proper

File: docs/proposals/keadm-debug.md [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+Usage:
+  keadm debug collect [flags]
+
+Examples:
+
+keadm debug collect --path
+
+Flags:
+  --config  Specify configuration file, defalut is /etc/kubeedge/config/edgecore.yaml
+  --path    Cache data and store data compression packages in a directory that defaults to the current directory
```
Comment (GsssC) [inline docs/proposals/keadm-debug.md:121]: how about change to `--output-path`?

File: docs/proposals/keadm-debug.md [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+  6. Use command `dns` Check whether the node domain name resolution function is normal
+  7. Use command `runtime `  Check whether the node container runtime function is installed, can use parameter `--runtime` to set container runtime,  default is docker
+  8. Use command `network `  check whether the node can communicate with the endpoint on the cloud,    can use parameter `--ip` to set test ip, default to ping clusterdns
+  9. Use command `pid ` check if the current number of processes in the environment is too many. If the number of available processes is less than 5%, the number of processes is considered insufficient
+
+`keadm debug collect`
+
+- What is it?
+
+  - This command will obtain all related data of the current node, and then locate and use  operation personnel
```
Comment (GsssC) [inline docs/proposals/keadm-debug.md:224]: determining the output file fomat , such as `.rar` or `.tar.gz`?

File: docs/proposals/keadm-debug.md (hunk 8 above)
Comment (shenkonghui) [reply docs/proposals/keadm-debug.md:186]: ok, I will removed this item

File: docs/proposals/keadm-debug.md (hunk 2 above)
Comment (shenkonghui) [reply docs/proposals/keadm-debug.md:53]: Obtain all the data of the current node, and provide it to the operation and maintenance personnel to locate the problem.  

Maybe translation error.

Comment (shenkonghui) [reply docs/proposals/keadm-debug.md:53]: > what dose it mean?

Obtain all the data of the current node, and provide it to the operation and maintenance personnel to locate the problem.

Maybe translation error.

File: docs/proposals/keadm-debug.md (hunk 3 above)
Comment (qingchen1203) [reply docs/proposals/keadm-debug.md:322]: Thank you for your suggestions for optimization，it will be modified soon

**Comment by Weaverzhu** [comment]
Update： now I use `corev1.List` to save all infos about resources like kubectl does, and json output is now the same as kubectl does, but yaml output is slightly different which is a bit annoying. Most of the implementation is copied from kubectl, but I'm using `yaml.Marshal` and `json.Marsal` for final output which is different  from kubectl. I'm currently trying to figure it out.

On the other hand, `kubeadm debug get <resourceType>` receive argument as `type` for querying db. So commands `./keadm debug get endpoints` or `./keadm debug get service` should get output as well.

Finally, as for namespaces and selectors, I found implementation in kubectl is a little bit heavy and I'm not sure if I can implement it in a short time. Any guide or help is welcomed!

Difference between `kubectl get pods -ojson` and `./keadm get pod`

```
root@edge:~/go/src/github.com/kubeedge/kubeedge/_output/local/bin# ./keadm debug get pod -ojson
{
        "apiVersion": "v1",
        "items": [
                {
                        "apiVersion": "v1",
                        "kind": "pod",
                        "metadata": {
                                "creationTimestamp": "2020-07-16T10:43:37Z",
                                "generateName": "kube-flannel-ds-amd64-",
                                "labels": {
                                        "app": "flannel",
                                        "controller-revision-hash": "6fbd6f4c89",

------------------split---------------


root@master:~# kubectl get pods -ojson
{
    "apiVersion": "v1",
    "items": [
        {
            "apiVersion": "v1",
            "kind": "Pod",
            "metadata": {
                "creationTimestamp": "2020-07-17T06:43:52Z",
                "generateName": "nginx-deployment-55cfcf74d-",
                "labels": {
                    "app": "nginx",
                    "pod-template-hash": "55cfcf74d"
                },
                "name": "nginx-deployment-55cfcf74d-rts8d",
                "namespace": "default",
                "ownerReferences": [
                    {
                        "apiVersion": "apps/v1",
```

current yaml output:

the term `object` shouldn't be in yaml output, which is different from kubectl, I currently trying to figure it out

```
root@edge:~/go/src/github.com/kubeedge/kubeedge/_output/local/bin# ./keadm debug get pod -oyaml
object:
  apiVersion: v1
  kind: List
  metadata:
    resourceVersion: ""
    selfLink: ""
items:
- object:
    apiVersion: v1
    kind: pod
    metadata:
      creationTimestamp: "2020-07-16T10:43:37Z"
      generateName: kube-flannel-ds-amd64-
      labels:
        app: flannel
        controller-revision-hash: 6fbd6f4c89
        pod-template-generation: "1"
        tier: node
      name: kube-flannel-ds-amd64-jxjlg
      namespace: kube-system
      ownerReferences:
      - apiVersion: apps/v1
        blockOwnerDeletion: true
```

**Pull request merged**